# SI 330 - Homework #1: Data Manipulation

## Background

This homework assignment focuses on the analysis of historical data from the Olympic games.  The description of the data includes the following:
> This is a historical dataset on the modern Olympic Games, including all the Games from Athens 1896 to Rio 2016. I scraped this data from www.sports-reference.com in May 2018.

Your main task in this assignment is to explore the data *using the data
manipulation methods we covered in class* as well as those in the assigned readings.  You may need to consult pandas documentation, Stack Overflow, or other online resources.  

** You should also feel free to ask questions on the [class Slack homework channel] (https://si330wn2019.slack.com/messages/CFA5AJPCL/)! **

A total of 50 points is available in this homework assignment, consisting of:
- 40 points for completing the specific homework questions. More comprehensive 
answers will tend to gain more points.
- 5 points for the overall quality of spelling, grammar, punctuation, and style of written responses.  (see https://faculty.washington.edu/heagerty/Courses/b572/public/StrunkWhite.pdf for a detailed specifications).
- 5 points for creating code that conforms to [PEP 8](https://www.python.org/dev/peps/pep-0008/) guidelines.  You should review those guidelines before proceding with the assignment.


## Answer the questions below. 
For each question, you should
1. Write code using Python and pandas that can help you answer the following questions, and
2. Explain your answers in plain English. You should use complete sentences that would be understood by an educated professional who is not necessarily a data scientist (like a product manager).

### <font color="magenta"> Q1 (6 points): Explore and Describe the dataset. 
- How many rows and columns are there in the data frame? Please use the appropriate dataframe property.
- Provide summary statistics (i.e. use the .describe() function) for age, height, and weight. Please present this information in a single dataframe.

uniqname = 'bpalko'

In [1]:
import pandas as pd
import numpy as np
athletes = pd.read_csv('data/athlete_events.csv')

In [2]:
athletes.shape

(271116, 15)

In [141]:
athletes[['Age', 'Height', 'Weight']].copy().describe()

Age         Height         Weight
count  261642.000000  210945.000000  208241.000000
mean       25.556898     175.338970      70.702393
std         6.393561      10.518462      14.348020
min        10.000000     127.000000      25.000000
25%        21.000000     168.000000      60.000000
50%        24.000000     175.000000      70.000000
75%        28.000000     183.000000      79.000000
max        97.000000     226.000000     214.000000

There are 15 columns in the data set. Each one describes a specfic aspect of an athlete. There are 271116 (rows) athletes in this dataset. The describe() function returns count, mean, std, min, 25%, 50%, 75%, and max for all athletes based on Age, Height, Weight.

### <font color="magenta"> Q2 (8 points): How many unique athletes are in the dataset? What proportion of unique athletes are female?
    - Ideally, your code should return a proportion without hardcoding.

In [146]:
athletes.ID.nunique()

135571

In [147]:
a_proportion = athletes.groupby('Sex').size()

In [148]:
a_proportion.loc['F']/(a_proportion.loc['F']+a_proportion.loc['M'])

0.2748712728131132

There are 135571 unique athletes. The proportion of females within the unique athletes is .2748.
 

### <font color="magenta"> Q3 (10 points): Looking at the time period from 1950 to today, which athelete competed in the most number of events? 
In which unique events did the athelete participate, and for what range of years? Which country did the athlete represent?

In [150]:
a_events = athletes[athletes.Year >= 1950].copy()
idx = a_events.groupby('ID').size().sort_values(ascending=False).head(2).index.tolist()


for ids in idx:
    print(a_events[a_events.ID == ids].Name.unique())
    print(a_events[a_events.ID == ids].Event.unique())
    print(a_events[a_events.ID == ids].NOC.unique())
    print(a_events[a_events.ID == ids].Year.min(), a_events[a_events.ID == ids].Year.max())
    print("\n")

['Andreas Wecker']
["Gymnastics Men's Individual All-Around"
 "Gymnastics Men's Team All-Around" "Gymnastics Men's Floor Exercise"
 "Gymnastics Men's Horse Vault" "Gymnastics Men's Parallel Bars"
 "Gymnastics Men's Horizontal Bar" "Gymnastics Men's Rings"
 "Gymnastics Men's Pommelled Horse"]
['GDR' 'GER']
1988 2000


['Takashi Ono']
["Gymnastics Men's Individual All-Around"
 "Gymnastics Men's Team All-Around" "Gymnastics Men's Floor Exercise"
 "Gymnastics Men's Horse Vault" "Gymnastics Men's Parallel Bars"
 "Gymnastics Men's Horizontal Bar" "Gymnastics Men's Rings"
 "Gymnastics Men's Pommelled Horse"]
['JPN']
1952 1964




Andreas Wrecker and Takashi Ono from 1950 to today competed in the most events. They competed in gymnastic events as such: Gymnastics Men's Individual All-Around, Gymnastics Men's Team All-Around, Gymnastics Men's Floor Exercise Gymnastics Men's Horse, Vault, Gymnastics Men's Parallel Bars, Gymnastics Men's Horizontal Bar, Gymnastics Men's Rings, Gymnastics Men's Pommelled Horse. Andreas competed for Geramany from 1988 to 2000. Takashi competed for Japan from 1952 to 1964.
 

### <font color="magenta"> Q4 (7 points): Which sport has the lowest median athlete age? Is there a tie?

In [149]:
ath_med=athletes.drop_duplicates(subset='Name').groupby('Sport').Age.median().sort_values().head()
ath_med

Sport
Rhythmic Gymnastics          18.0
Swimming                     19.0
Figure Skating               20.0
Gymnastics                   21.0
Short Track Speed Skating    21.0
Name: Age, dtype: float64

The lowest median age is in "Rhythmic Gymnastics" and the age is 18.

### <font color="magenta"> Q5 (8 points): In which year did Canada win the most medals?

In [15]:
a5 = athletes.copy()

In [151]:
a5 = athletes.dropna(subset= ['Medal'])
a_country = a5[a5.NOC == 'CAN']
a_country.Year.value_counts().sort_values(ascending=False).head(1)

1984    91
Name: Year, dtype: int64

In 1984, Canada had its most medals earned.

### <font color="magenta"> Q6 (11 points): In which sports in the 2014 Winter Olympics did female medalists win more medal points than male medalists?
    3 points are assigned for a gold medal
    2 points for silver
    1 point for bronze

In [21]:
a6 = athletes.copy()

In [22]:
def label_race (a6_seas_win_d):
    if a6_seas_win_d['Medal'] == 'Bronze' :
        return 1
    elif a6_seas_win_d['Medal'] == 'Silver' :
        return 2
    else :
        return 3

In [23]:
a6['Points']=a6.apply (lambda row: label_race (row),axis=1)
a6d=a6.dropna(subset=['Medal'])

In [152]:
a6_seas = a6d[a6d.Season == 'Winter']
a6_seas_win= a6_seas[a6_seas.Year == 2014]
a6_sum = a6_seas_win.groupby(['Sport','Sex']).Points.sum()
a6_sum = a6_sum.unstack()
a6_female = a6_sum[a6_sum['F'] > a6_sum['M']]

a6_female

Sex                           F     M
Sport                                
Short Track Speed Skating  45.0  42.0
Speed Skating              54.0  48.0

Females during the 2014 Winter Olympics won more medal points in Short Track Speed Skating, and Speed Skating. For Short Track Speed Skating, Females had 45 to Men's 42. In Speed Skating, Females had 54 to Men's 48.

### QBonus 5 pt: For each year in which games were held, what proportion of the host country medalists were women? 
You will need to combine multiple datasets to complete the analysis.
<p>Suggested data: 
<p>- https://en.wikipedia.org/wiki/List_of_Olympic_Games_host_cities (note: we suggest you use pd.read_html)

In [ ]:
# put your code here

(Use this space to explain your answers)

## Please submit your completed notebook in .IPYNB and .HTML formats via Canvas